### ISSUE TITLE
Add NoiseGenerator Node

### ISSUE THREAD

**Issue by g200kg** [WebAudio/web-audio-api-v2#8]
current possible choice may be:
 1) Javascript code with ScriptProcessor -- ScriptProcessor is deprecaed.
 2) AudioWorker -- Not yet sure, but It seems slightly elaborated for these simple function.
 3) BufferSourceNode with pre-rendered buffer -- may be ok but wastetfull?
I think it is a bit cumbersome relative to other basic waveform.

hopefully, white / pink switchable is great.

**Comment by rtoy** [comment]
Some additional notes that came up during the TPAC discussion.  This is a summary.

This can be polyfilled with an AudioWorklet, but won't work in a way that I ( @rtoy) would like:

- No requirement that each browser uses the same method for `Math.random()`
- The seed for the RNG can't be set, so you can't get consistent results for each run (if desired).  but Ruth John pointed out that TC39 is considering adding support for specifying a seed
- There's no way to guarantee independent sequences for each generator.  Just specifying a different seed for each generator doesn't guarantee the sequences are not related.

By defining built-in node, we can specify exactly how the generator behaves and provide a seed and a means of getting guaranteed independent sequences for each node.

@svgeesus also pointed out that as designed here, the noise sequence isn't bandlimited.  That's true, but I think this is what we want.  The power spectrum of this signal will be flat as desired.

**Comment by svgeesus** [comment]
[seedable PRNG from Web Crypto](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/getRandomValues)

**Comment by padenot** [comment]
Virtual F2F:
- `pcg` seems compelling, it has all the property we need, is very compact in terms of code. The source code license is Apache so it's generally ok
- The author provides guidance for converting to floating points correctly, in [0.0, 1.0], we'd scale to (-1.0; 1.0], which has a bias towards +1.0, as float32
- Being able to have multiple uncorrelated generators from the same seed is useful
- All this might be overkill, so we'll sanity check with people who know about this
- There is precedent to use `TypedArrays` in the interface of random generator code, this might be a good interface choice here (to avoid the issues with JS's double/ `Numbers.MAX_SAFE_INTEGER`, etc.)

**Comment by rtoy** [comment]
Updated proposal:
```
interface NoiseGeneratorNode {
  constructor (BaseAudioContext context, NoiseGeneratorOptions options);
  readonly attribute unsigned long sequenceNumber;
  // seed must have length >= 2, in the same order as initialSeed
  // below.  It's an error if it's too short.
  void getSeed(sequence<unsigned long> seed);
};

dictionary NoiseGeneratorOptions {
  NoiseType type = "uniform";
  required unsigned long sequenceNumber;
  // Must have length >= 2. It's an error if it's too
  // short. initialSeed[0] is the low 32 bits of the 64-bit seed.  If
  // initialSeed is not given, the browser creates a random initial
  // seed in an unspecified manner.
  sequence<unsigned long> initialSeed;
};
```

With this information, you can recreate a generator that would produce
the same sequence of samples.

We have, however, dropped the ability to examine the current state.
Since the state is updated on the audio thread, it's probably not all
that useful to get the state in the main thread.  But this also means
you can't suspend the context, get the state, and create a new
generator that would pick up exactly where you left off.  I don't see
that as a major deal-breaker here.

**Comment by guest271314** [comment]
One approach to create a noise generator is to utilize `OscillatorNode`s


```
let ac = new AudioContext();
let generate = {on: true};
async function* noise(g, frequencyRange = [50, 880]) {
  const [min, max] = frequencyRange;
  const types = ['sine', 'square', 'sawtooth', 'triangle'];
  while (g.on) {
    yield new Promise(resolve => {
      const type = types[Math.floor(Math.random() * types.length)];
      const osc = new OscillatorNode(ac, {
         type, 
         frequency:Math.floor(Math.random() * (max - min + 1) + min)
       });
      osc.onended = _ => {
       osc.disconnect();
       resolve();
      }
      osc.connect(ac.destination);
      osc.start(ac.currentTime);
      osc.stop(ac.currentTime + .05);
    })
  }
}

(async() => {
  for await (const _ of noise(generate, [25, 100])) {
    if (ac.currentTime > 30) generate.on = false;
  }
  await ac.close();
  console.log('done');
})();


```

**Comment by svgeesus** [comment]
> We have, however, dropped the ability to examine the current state.

Does that mean that sample-and-hold on noise is no longer possible? I mean for a one-shot its easy to make another pseudo-random generator, I was thinking more of S/H on noise with lag, (standard ARP 2600 or Odyssey effect)

**Comment by rtoy** [comment]
Previously in https://github.com/WebAudio/web-audio-api-v2/issues/8#issuecomment-519621343, the node had
```
readonly attribute unsigned long long state;
```
This allowed you to examine the internal state of the generator.   You could get the internal state, but it wasn't useful because we didn't provide a method for you to update the internal state of a generator.  So I removed it.  

This means you couldn't inspect a noise node and create a new node that would start off with exactly the same values as the original node.  But that wasn't really feasible anyway unless you stopped the context to get a stable value of the state and initialized the new node with that state. (But you couldn't create the node with that state; you only have the seed which is used with the sequence number to create the state.)

Is that important to anyone?

And then, do you want to node to do S/H of the values?  That's not currently part of the proposed spec.  We can add that.

**Comment by rtoy** [comment]
@guest271314 Your example in https://github.com/WebAudio/web-audio-api-v2/issues/8#issuecomment-710193331 doesn't generate white noise AFAICT.  Looks like it randomly selects one of the oscillator types and then a random frequency. That's not white noise, which is what we want here.

**Comment by guest271314** [comment]
The term "white noise" can have various meanings and really should not be used here. Since "race" theory is still officially practiced and is policy in the United States per O.M.B. Directive 15, individuals and institutions in the United States that are intellectually lazy and do not bother to read the definitions, tend to associate color with the fictitious political classifications "black" and "white", for example, see [In contemporary use, does 'white noise' enjoy common figurative meaning?](https://english.stackexchange.com/q/488005); [White Noises, Woman or Women](https://english.stackexchange.com/q/487542); and per [Issue 842296: Use inclusive terminology throughout Chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=842296) the term "white noise" would probably be a recommendation for substitution; substituting "allowlist" for "whitelist" throughout Chromium source code, e.g., https://bugs.chromium.org/p/chromium/issues/detail?id=1142696, and the problems (see this post https://github.com/guest271314/banned/issues/2#issuecomment-711082277 at [GitHub censors the term "Negro"](https://github.com/guest271314/banned/issues/2, where "Negro" is censored by GitHub but not "White" though both are equally fictitious political classifications within "race" theory) and linked articles I published that no individual or institution has been able to, nor will ever be able to repudiate, as I am an expert in the subject matter of "race" theory) that arise when organizations attempt to arbitrarily censor terms, again, due to, in general, individuals' laziness, failure to read and cite primary source definitions, and reliance on European folklore rather than actual historical facts.

Even in this very issue there are various proposals for what the term "white noise" means to individuals, yet none cite an actual primary source definition. So, that needs to be clearly defined first, yet to avoid any potential for ambiguity re the theory of a "white" "race" I would strongly recommend abandoning associating a color with a noise generator, as color frequencies and sound frequencies are not 1:1 whatsoever.

**Comment by guest271314** [comment]
It must be noted here that within the domain of primary source research Wikipedia is at best a secondary source, and cannot be cited for any primacy. 

Why not name the method "black noise" generator? Or a "yellow noise" generator? How would that sound?

From perspective here a "white noise" generator in code `#fffff` would be silence, not sound.

You can see the problem with attempting to associate color with sound. Color visible to humans, in general is 400nm to 800nm, blue to red, with green being the color human eyes detect and focus on more than other colors - though even within that scope there are still individuals that are color blind or are vision impaired: How does "white noise" relate to them? 

As an aside I will not here that of the thousands of individuals and institutions that I have asked: "What primary source definition of the term 'white' are you relying on?" and "What primary source definition of the term 'black' are you relying on?" when they use the terms in reference to natural persons, only one (1) has actually stated that they are relying on the same definition that the U.S. Census Bureau uses. The vast majority of individuals infer that "white" "race" or "black" "race" are based on "color", or "skin" "color", or that they can determine if an individual is "black" "race" or "white" "race" with their own eyes. The problem is that the official defintion of "race" and "black" "race" and "white" "race" is not based on science whatsoever - and the terms "color" and "skin" are not found in the official U.S. administratiove regulation definition at all. So, for the above stated reasons, this noise generator should not be associated with any color at all. If users in the field want to make such claims then they can do so on their own, and will quickly find that, as stated above the

**Comment by rtoy** [comment]
White noise is a random process with a constant power spectral density.

**Comment by guest271314** [comment]
If we just read Wikipedia linked at https://github.com/WebAudio/web-audio-api-v2/issues/8#issuecomment-173689784, which I prefer to not quote, as again, Wikipedia is not a primary source for anything, we find

> **Black noise**
>
> - Silence
> - ...
> - Noise that has a frequency spectrum of predominantly zero power level over all frequencies except for a few narrow bands or spikes. Note: An example of black noise in a facsimile transmission system is the spectrum that might be obtained when scanning a black area in which there are a few random white spots. Thus, in the time domain, a few random pulses occur while scanning.

essentially carrying out western academic biases with regard to the political classes created by itself, without compunction, associating "black noise" with "zero power", akin to "whitelist" and "blacklist", the premise being that "white" is somehow powerful while "black" is somehow powerless. Subtle yet consistent across western academia; and in fact, can and has been a tool used to manipulate individuals minds and culture long-term - for indivudals that are not capable of rejecting spurrious claims on their face - no matter where the claim is derived from.

From an [answer](https://sound.stackexchange.com/a/3034) at [What do colors sound like?](https://sound.stackexchange.com/questions/3029/what-do-colors-sound-like)

> Yes, you can convert color to sound.
> 
> No, orange doesn't have a particular frequency or sound. You can pick whatever mapping you want. Do you want orange be a specific sine frequency, or should it represent a square wave? You pick.

**Comment by guest271314** [comment]
> White noise is a random process with a constant power spectral density.

Per what?

**Comment by guest271314** [comment]
What is "Black noise", and again, per what?

There is no constant mapping of colors to sound. And we should not be making that claim here. It is not necessary. What is necessary is clearly defining what the expected output is; random, consistent; both?

**Comment by guest271314** [comment]
> a random process with a constant power spectral density.

Is perfectly capable of standing on its own without prepending European folklore to the techinical term.

**Comment by rtoy** [comment]
> > White noise is a random process with a constant power spectral density.
> 
> Per what?

If Wikipedia is not definitive for you, here are two references I have that define white noise exactly this way:

- Papoulis, _Probability, Random Variables, and Stochastic Processes_, a classic book on probability
- Proakis, _Digital Communications_, another classic book on digital communications

**Comment by rtoy** [comment]
Oh, I didn't know Javascript has BigInt now.  And Chrome Beta and Firefox Developer support it.  This would make for a nicer API.  But it looks like WebIDL doesn't currently support BigInt, so perhaps we're stuck with using arrays to specify the seed.

**Comment by rtoy** [comment]
Anyway, here is a proposal assuming we had BigInt:
```
enum NoiseType {
  "uniform",
  "gaussian" // or "normal"?
};

interface NoiseGeneratorNode {
  constructor (BaseAudioContext context, NoiseGeneratorOptions options);
  // Different sequenceNumbers with the same seed produces independent
  // sequences of random numbers.
  readonly attribute unsigned long sequenceNumber;
  // The 64-bit seed used to initialize this generator.
  readonly attribute BigInt seed;
};

dictionary NoiseGeneratorOptions {
  NoiseType type = "uniform";
  required unsigned long sequenceNumber;
  // 64-bit seed.  If not given, the browser creates a random initial
  // seed in an unspecified manner.
  BigInt seed;
};
```

**Comment by domenic** [comment]
Some naive questions trying to help out on the BigInt side. I tried to answer these by reading upthread but I couldn't find them easily, probably because this is mostly domain experts talking and goes above my head :).

- Do you definitely need 64 bits for an initial seed? It's not clear to me why that's better than 32 or 53.

- Why is it preferable to get your 64 bits in integer form, instead of in double form? I.e. why not take a `double` as input, and just use its bit pattern (which is 64 full bits)? You could even reinterpret those bits as an 64-bit integer if it made calculations easier (example below)

- If you go with BigInt, what is your plan for handling values outside the desired range? (Throwing seems perfectly reasonable, FWIW.) Is your desired range ~ [-2^63, 2^63] or [0, 2^64]? (I know there should be +/-1s in some of those endpoints... I always forget which.)

To be clear, I am completely ready to believe this is best done with 64-bit ints, and am excited about BigInt getting its first web platform use case. I just want to make sure we have all our reasoning lined up, so that we can set a nice precedent for the future.

<details>
<summary>Code to reinterpret a 64-bit float as a 64-bit int</summary>

```js
const buffer = new ArrayBuffer(8);
const intView = new BigInt64Array(buffer);
const floatView = new Float64Array(buffer);

floatView[0] = Math.PI;
console.log(intView[0]);
```
</details>

**Comment by guest271314** [comment]
The links are merely to authors that created arbitrary terms of art for the processes they were theorizing about

> **_White noise_** We shall say that a process V(I) is white noise if


>  The output, called _filtered white noise_, is denoted by X(t). This process
has a power spectral density of


The term "White Noise" within the scope of sound is not based on science, rather on euphemism, or, frankly, just nicknames created from thin air by the attributions, contrary to the actual physics.

I reject the claims in those documents relevant to associating or assigning "white noise" to have some supposed consistent power while "black noise" is characterized as "silence" or without power - yet nothing could be further from the truth - in any domain save for the laboratories of the individuals who invented the terms. 

I directly reject such coinage as inconsistent with the rest of the physical world outside of their laboratories. 

This specification must not adopt such continuace of mere conjecture - when it is not necessary: use terms other than colors for noise generation: A very simple substitution is "_static_" of varying ranges. 

See this [answer](https://physics.stackexchange.com/a/159862) at [Why is black the best emitter?](https://physics.stackexchange.com/q/159856)

> **And now we answer the question**
>
> A material which is a very good absorber looks black because it absorbs all incoming light. Well that sounds convincing but if you think about it you would point out that I told you that good absorbers are also good emitters, so a good absorber should give off light too. It shouldn't look black! The key here is that it doesn't have to emit the photons at the same wavelength as it absorbs them. What generally happens is that after an e.g. visible photon comes in and gets absorbed, the energy is converted to heat, and it is later emitted by a thermal emission as described above. This thermal emission may be in the infra-red range (for practical temperatures), which you can't see with your eyes. That's why an object which is both a good absorber and a good emitter looks black when it's at thermal equilibrium near room temperature.[d]
>  Note that this means that black objects are actually radiating more power than white ones, you just can't easily tell because that power is in a wavelength you don't see.

Pay careful attention the the last paragraph, above. 

Coining of phrases without a real connection to science is just hearsay, just because a developer decided to name their process a color - when they did not have to - and we do not have to here.

> **More information**
> A hypothetical object which perfectly absorbs all incoming radiation is called a black body. Using quantum mechanics and statistical mechanics you can compute the amount of radiation power a black body at a given temperature should emit at each wavelength.[b]
>  Interestingly, if you look at the plot of the sun's emission, you see that it's pretty close to an ideal black body.

Therefore it is physically impossible for "white" "noise" to absord, maintain, or emit as much power as "black", specifcally an ideal black body. 

There is no such thing as an ideal "white" body or "white noise" that in any theory that comes anywhere close to the power of black body radiation.

As I stated eariler, "sound" and "color" are not 1:1, scientifically; [Neil deGrasse Tyson explains why dark matter matters (and is kind of our frenemy)](https://www.popsci.com/why-dark-matter-matters/)

> As a wave, light was thought to require a medium through which to propagate its energy, much as sound requires air or some other substance to transmit its waves. But light turns out to be quite happy traveling through the vacuum of space, devoid of any medium to carry it. Unlike sound waves, which consist of air vibrations, light waves were found to be self-propagating packets of energy requiring no assistance at all.

Note the differentiation between sound waves and light waves. The phenomenon are of different types. 

If anything "black noise" is the drum, which was banned in the prisoner of war camps termed "Several States"; the bassline; the blues; the 808; the kick drum; Jazz; breaking out dancing in the isles at the supermarket when that song comes on over the store sound system - while everybody else walks by in "silence" not knowing what the hell is going on 'cause they really 'aint emitting no energy at all; the origin of music itself, Ancient African Egyptians. That is, if the standard is just what an individual names their function, and abandoning the scientific method to substantiate there is actually a such thing as "white noise" outside of the documents which created the terms. 

The science is clear. If individuals attempt to turn science and facts on their head I turn their spurrious claims on their heads - until they ban me - but can never answer the questions or repudiate the facts stated.

That is, given the stanard is just people naming their processes "white noise" and assigning them "power", "black noise", in this specification, must, becuase that is the physical fact, emit more power - at all levels - than "white noise" in sound output.

**Comment by rtoy** [comment]
The algorithms used by the pcg generator specify a 64-bit seed.  Using only 32 bits would restrict the the set of possible sequences, or make the sequences less random.  I don't know how to analyze such things.  (Emails to the author have been unanswered.)  53 bits might be enough, but again I don't know.  (We previously allowed you to examine the state of the generator of 64 bits, and all of those are needed exactly as is.)

Yes, using a double and treating it as a 64-bit integer could  work too.  But there's  a huge range of values that are NaNs that would be hard to get.  To get the full range, you'd basically do the opposite of your code to reinterpret doubles to 64-bit ints.  This would be a rather unnatural interface.

For BigInts, we'd probably just take the low 64 bits to get the seed we need.

BigInts provide a reasonably natural interface that reflects  the pcg describes.

Without BigInts, I think the best interface would be an array of 2 32-bit integers (or any other integer array that has 64 bits total).

In conclusion, BigInts provide a natural interface, but it's not strictly required.  An array of integers works fine.  Using fewer than 64 bits has unknown effects on the quality of the generated output.

**Comment by rtoy** [comment]
> The links are merely to authors that created arbitrary terms of art for the processes they were theorizing about
Fundamentally, all terms are arbitrary.

We understand your point. Please refrain from further comments.  We get it.

**Comment by rtoy** [comment]
Since https://github.com/heycam/webidl/pull/525 has been merged, perhaps we should use BigInt's for the interface.  It may take some time before browsers get around to supporting this WebIDL feature though.

Not sure how we want to deal with that.

**Comment by rtoy** [comment]
Teleconf:  @padenot will ask SpiderMonkey folks about exposing this on the web.

@rtoy is perfectly happy with using arrays.  We can add support for bigint later if desired.

**Comment by rtoy** [comment]
Teleconf:  Generally agreed not use bigints in the interface; an array is good enough.

We can add it later if desired.

**Comment by rtoy** [comment]
F2F meeting:  Agreed that the API is correct  except BigInt should not be used.  Michel Buffa will look at some Faust code and possibly propose some additional noise types if they are recognized as good methods for generating such noise.

**Comment by rtoy** [comment]
Here's the updated API using arrays:
```
enum NoiseType {
  "uniform",
  "gaussian"  // Or "Gaussian" because it's a name?
};

interface NoiseGeneratorNode : AudioScheduledSourceNode {
  [RaisesException] constructor(BaseAudioContext context,
                                optional NoiseGeneratorOptions options = {});
  attribute NoiseType type;
  readonly attribute unsigned long sequenceNumber;
  [RaisesException] void getSeed(Uint32Array seed);
};

dictionary NoiseGeneratorOptions {
  NoiseType type = "uniform";
  unsigned long sequenceNumber = 1;
  sequence<unsigned long> initialSeed;
};

```

The RNG to use is the minimal C (reference) implementation: https://github.com/imneme/pcg-c-basic.  This is small and fast and passes all the tests for randomness.  It is specifically not crypto safe.

`initialSeed` is a sequence with a length of at least two.

The 64-bit seed, `rngSeed`, for the underlying generator is computed from `initialSeed` by

```
rngSeed = (initialSeed[0] << 32) + initialSeed[1]
```

That is, the first element of `initialSeed` is the high 32-bits of the seed and the second element is the low 32-bits of the seed.  Any other elements are ignored.


`getSeed` must throw an error if the length of the array `seed` is less than two.

The array `seed` is filled according to
```
seed[0] = (rngSeed >> 32) & 0xffffffff;
seed[1] = rngSeed & 0xffffffff;
```

If `initialSeed` is not provided, the UA must generate a random seed in a UA-specific manner such that creating multiple `NoiseGeneratorNode`s will all have a different seed.  Thus:

```
let n1 = new NoiseGeneratorNode(context, {sequenceNumber: 1});
let n2 = new NoiseGeneratorNode(context, {sequenceNumber: 2});
let s1 = new Uint32Array(2);
let s2 = new Uint32Array(2);
n1.getSeed(s1);
n2.getSeed(s2);

// Must return false
s1.every((val, index) => val !== s2[index])
```

Conversely, two `NoiseGeneratorNode`s with the same `sequenceNumber` and `initialSeed` must produce exactly the same output.


For any given seed and sequence number, every UA must produce **exactly** the sequence of numbers when the type is "uniform".  For "gaussian", this is not required; UAs are free to choose any appropriate algorithm to generate Gaussian noise.
